packages/core/src/api/apis/definitions/featureFlags.ts

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import ApiRef from '../ApiRef';
+import { FeatureFlagName } from '../../plugin/types';
+
+/**
+ * The feature flags API is used to toggle functionality to users across plugins and Backstage.
+ *
+ * Plugins can use this API to register feature flags that they have available
+ * for users to enable/disable, and this API will centralize the current user's
+ * state of which feature flags they would like to enable.
+ *
+ * This is ideal for Backstage plugins, as well as your own App, to trial incomplete
+ * or unstable upcoming features. Although there will be a common interface for users
+ * to enable and disable feature flags, this API acts as another way to enable/disable.
+ */
+
+export enum FeatureFlagState {
+  NotEnabled = 0,
+  Enabled = 1,
+}
+
+export type FeatureFlagsApi = {
+  /**
+   * Get a list of the user's currently enabled feature flags.
+   * Reads directly from window.localStorage Browser API.
+   *
+   * @returns string[] enabledFeatureFlags List of feature flags enabled by the current user
+   */
+  getEnabledFeatureFlags(): Set<FeatureFlagName>;
+
+  /**
+   * Check the feature flag name convention.
+   * Used in the `registerFeatureFlag` method as well as in the `set` method.
+   *
+   * @returns string[] errors List of errors as string. Empty array if no errors.
+   */
+  checkFeatureFlagNameErrors(name: FeatureFlagName): string[];
+
+  /**
+   * Use Feature Flags as a React Hook.
+   */
+  useFeatureFlag(
```

Comment:
Have been searching for a good solution for how to combine hooks and the API concept, but I don't think this is it. For now I'd rather keep hooks separate from the API and as thin as possible. I think the consumption pattern if this is alright, but tbh I think it would be even better to just have a `useFeatureFlag` hook that uses `useApi` internally to grab the API and read the flag.

We also want to avoid react concepts in these APIs as much as possible. The idea is that the APIs should be usable from any frontend framework.

Reply:
> We also want to avoid react concepts in these APIs as much as possible. The idea is that the APIs should be usable from any frontend framework.

That sounds very strange to me. Unless we have plans to migrate from React, why wouldn't we want the APIs to be as idiomatic as possible? Maybe I'm completely missing the point, but I completely see this more as a React FeatureFlag library than a JavaScript FeatureFlag library. Can you share more on why Backstage thinks we should build APIs without React in mind?

I like the idea though of abstracting out the `useApi` although I wanted to refrain from that and instead try to be as consistent as possible with other APIs (aka `ErrorApi`).

Reply:
Apart from the possibility to migrate to other frameworks, it's because react has it's own state and dependency model. Given this API, how would another API consume this API to check if a feature flag is enabled? It would have to in turn be exposing it's API as a hook and consume this hook. For many APIs that's going to be a PITA to manage, and will also cause performance trouble because of rerenders being triggered when there's really no need.

What we've gone with internally so far is to stay away from react in the implementation of APIs, and then have a minimal bridge to react at the "edge" when it's needed.

I also want to flip this question around, what benefit do you get from implementing APIs using react concepts?

Reply:
The main benefit I was thinking of was one interface - a React interface. I understand it acts as a view layer, but hooks (even if you don't use underlying React concepts such as state and hooks) operate as good mental models for designing complex web applications like Backstage. Despite `const [getter, setter] = useFeatureFlag('feature name');` is a React pattern, it's quite nice to work with.

If the plan is for other plugins to use it as an API, then I can see how it might make sense. Some Backstage plugins might not use React at all, but still might want to use certain APIs. The obvious case in point might be a networking or caching layer. In that case, I take your point and agree with it :D 

Nonetheless, I have removed all the React-specific logic and now more simple as per your comments!